// @gogovega/node-red-contrib-firebase-realtime-database/build/lib/types/DatabaseNodeType.js
"use strict";
/**
 * Copyright 2022-2023 Gauthier Dandele
 *
 * Licensed under the MIT License,
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/MIT.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionStatus = void 0;
var ConnectionStatus;
(function (ConnectionStatus) {
    ConnectionStatus[ConnectionStatus["DISCONNECTED"] = 0] = "DISCONNECTED";
    ConnectionStatus[ConnectionStatus["CONNECTING"] = 1] = "CONNECTING";
    ConnectionStatus[ConnectionStatus["CONNECTED"] = 2] = "CONNECTED";
    ConnectionStatus[ConnectionStatus["LOG_OUT"] = 3] = "LOG_OUT";
    ConnectionStatus[ConnectionStatus["NO_NETWORK"] = 4] = "NO_NETWORK";
    ConnectionStatus[ConnectionStatus["ERROR"] = 5] = "ERROR";
})(ConnectionStatus || (exports.ConnectionStatus = ConnectionStatus = {}));
