// @gogovega/node-red-contrib-firebase-realtime-database/build/lib/types/FirebaseNodeType.js
"use strict";
/**
 * Copyright 2022-2023 Gauthier Dandele
 *
 * Licensed under the MIT License,
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/MIT.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryConstraint = void 0;
var QueryConstraint;
(function (QueryConstraint) {
    QueryConstraint[QueryConstraint["endAt"] = 0] = "endAt";
    QueryConstraint[QueryConstraint["endBefore"] = 1] = "endBefore";
    QueryConstraint[QueryConstraint["equalTo"] = 2] = "equalTo";
    QueryConstraint[QueryConstraint["limitToFirst"] = 3] = "limitToFirst";
    QueryConstraint[QueryConstraint["limitToLast"] = 4] = "limitToLast";
    QueryConstraint[QueryConstraint["orderByChild"] = 5] = "orderByChild";
    QueryConstraint[QueryConstraint["orderByKey"] = 6] = "orderByKey";
    QueryConstraint[QueryConstraint["orderByPriority"] = 7] = "orderByPriority";
    QueryConstraint[QueryConstraint["orderByValue"] = 8] = "orderByValue";
    QueryConstraint[QueryConstraint["startAfter"] = 9] = "startAfter";
    QueryConstraint[QueryConstraint["startAt"] = 10] = "startAt";
})(QueryConstraint || (exports.QueryConstraint = QueryConstraint = {}));
