// @gogovega/node-red-contrib-firebase-realtime-database/build/lib/types/FirebaseConfigType.js
"use strict";
/**
 * Copyright 2022-2023 Gauthier Dandele
 *
 * Licensed under the MIT License,
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/MIT.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = exports.Listener = void 0;
var Listener;
(function (Listener) {
    Listener["value"] = "onValue";
    Listener["child_added"] = "onChildAdded";
    Listener["child_changed"] = "onChildChanged";
    Listener["child_moved"] = "onChildMoved";
    Listener["child_removed"] = "onChildRemoved";
})(Listener || (exports.Listener = Listener = {}));
var Query;
(function (Query) {
    Query[Query["set"] = 0] = "set";
    Query[Query["push"] = 1] = "push";
    Query[Query["update"] = 2] = "update";
    Query[Query["remove"] = 3] = "remove";
    Query[Query["setPriority"] = 4] = "setPriority";
    Query[Query["setWithPriority"] = 5] = "setWithPriority";
})(Query || (exports.Query = Query = {}));
