// @gogovega/node-red-contrib-firebase-realtime-database/build/lib/types/OnDisconnectConfigType.js
"use strict";
/**
 * Copyright 2022-2023 Gauthier Dandele
 *
 * Licensed under the MIT License,
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/MIT.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = void 0;
var Query;
(function (Query) {
    Query[Query["none"] = 0] = "none";
    Query[Query["cancel"] = 1] = "cancel";
    Query[Query["set"] = 2] = "set";
    Query[Query["update"] = 3] = "update";
    Query[Query["remove"] = 4] = "remove";
    Query[Query["setWithPriority"] = 5] = "setWithPriority";
})(Query || (exports.Query = Query = {}));
