// @gogovega/node-red-contrib-firebase-realtime-database/build/nodes/database.js
"use strict";
/**
 * Copyright 2022-2023 Gauthier Dandele
 *
 * Licensed under the MIT License,
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/MIT.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const databaseNode_1 = __importDefault(require("../lib/databaseNode"));
module.exports = function (RED) {
    function DatabaseNode(config) {
        RED.nodes.createNode(this, config);
        const self = this;
        self.connectionStatus = 0;
        self.config = config;
        self.nodes = [];
        self.RED = RED;
        const database = new databaseNode_1.default(self);
        database.logIn();
        self.on("close", (done) => database
            .logOut()
            .then(() => done())
            .catch((error) => done(error)));
    }
    RED.nodes.registerType("database-config", DatabaseNode, {
        credentials: {
            apiKey: { type: "text" },
            clientEmail: { type: "password" },
            email: { type: "text" },
            json: { type: "password" },
            password: { type: "password" },
            privateKey: { type: "password" },
            uid: { type: "text" },
            url: { type: "text" },
        },
    });
};
